Jia multi gpu eval

liyongsea opened the pull request:
# What does this PR do?
This PR provdie a script to run codeparrot human eval in multi GPU mode
The code is a enhanced version of human_eval.py

# Performance data
The performance is check using the following command
```
accelerate launch scripts/human_eval.py --model_ckpt lvwerra/codeparrot-small --do_sample True --temperature 0.2 --top_p 0.95 --n_samples=40 --HF_ALLOW_CODE_EVAL="1" --device_int=0 --num_tasks=8 --batch_size=10
```
One should obtain something similar to 
```
Results: {'pass@1': 0.1375, 'pass@10': 0.21251641317430792}
```

# Time benchmark
First configure accelerate
```
accelerate config
```

Remember to put more than 1 process if want to leverage multip gpu.
Then run
```
accelerate launch scripts/human_eval.py --model_ckpt lvwerra/codeparrot-small --do_sample True --temperature 0.2 --top_p 0.95 --n_samples=200 --HF_ALLOW_CODE_EVAL="1" --batch_size=10
```
One should obtain something similar to 
```
Results: {'pass@1': 0.03807926829268292, 'pass@10': 0.056158491636524525, 'pass@100': 0.0739473539338153}                                                                                                  
```

Here is time performance benchmark under a GPU VM with 4xT4
| Number of processes | Evaluation time for lvwerra/codeparrot-small | 
|---------------------|----------------------------------------------|
| 1                   | 2:10:00                                      |  
| 4                   | 35:00                                        | 

You could also test the large model by setting `--model_ckpt lvwerra/codeparrot`, you might probably need to set a small batch size (1 is usually adapted)
Warning: when testing with the large model, accelerate is loading the model on all the processes. You would need about 60G RAM on your VM. (I am not sure there is another alternative)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

Commit:
add simple multi gpu complet

Commit:
add human_eval_multi_gpu

Commit:
use copy strategy to distribute across gpu, to avoid padding

Commit:
add doc string

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_16428). All of your documentation changes will be reflected on that endpoint.

LysandreJik commented:
Hey @liyongsea, thanks for your PR! Could you rebase/merge on `main` so that the CI passes? Thank you!

lvwerra commented:
There seems to be also a minor issue with the code style. The CI enforces some standards and you can fix any formatting issues with:

```bash
make fixup
```

For more information on both rebasing/merging and quality checks you can checkout the [contributing guide](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md).

lvwerra commented on examples/research_projects/codeparrot/scripts/human_eval_mgpu.py:
```diff
@@ -0,0 +1,218 @@
+    where nc is the number of copies of the prompt, and nt is the number of tasks.
+    nc is such that num_sample = nc * batch_size
+    In this way, the same prompt is sent to the model on all the GPUs to avoid zero padding.
+    (accelerate require all the batch having the same shape before feeding it to multiple GPUs)
```
Do I understand this correctly: if we want 100 generations and have 4 GPUs then each GPU will get 25 times the same sample. However, if we have 8 GPUs this will break, as 100 is not divisible by 8, right? As we scale to more (and larger) models I suspect that we will likely scale the GPUs and so that would be quite a limiting factor.

liyongsea replied:
Yes, this is the idea

lvwerra commented on examples/research_projects/codeparrot/scripts/human_eval_mgpu.py:
```diff
@@ -0,0 +1,218 @@
+            # without strip, the model generate commented codes ...
+            prompt = self.tokenizer.eos_token + self.dataset[task]['prompt'].strip()
+            # codeparrot model is not robust to padding
+            input_ids = self.tokenizer(prompt)['input_ids']
```
We could still pad here and remove the paddings before generation to play nicely with both `accelerate` and the generation. See my comment below.

liyongsea replied:
Good idea ! I haven't though about this

lvwerra commented on examples/research_projects/codeparrot/scripts/human_eval_mgpu.py:
```diff
@@ -0,0 +1,218 @@
+    for step, batch in tqdm(enumerate(dataloader)):
+        with torch.no_grad():
+            gen_kwargs["stopping_criteria"][0].start_length = batch["ids"].shape[-1]
+            generated_tokens = accelerator.unwrap_model(model).generate(
```
If we apply padding to the inputs for accelerate distribution we could remove it here just before the generation step. At this point the tensor should already be distributed and padding no longer necessary.

lvwerra commented on examples/research_projects/codeparrot/scripts/human_eval_mgpu.py:
```diff
@@ -0,0 +1,218 @@
+            # codeparrot model is not robust to padding
+            input_ids = self.tokenizer(prompt)['input_ids']
+            for _ in range(self.n_copies):
+                yield {"ids": torch.tensor(input_ids)}
```
If we move to the more general way to distribute the task with the padding, it might become harder to track which sample belongs to which task. We could also yield the `task_id` so that after the generation we can easily match all the generations with the tasks again.

lvwerra commented on examples/research_projects/codeparrot/scripts/human_eval_mgpu.py:
```diff
@@ -0,0 +1,218 @@
+
+    n_tasks = args.num_tasks if args.num_tasks is not None else len(human_eval["test"])
+    n_copies = args.n_samples // args.batch_size
+    if n_copies % accelerator.num_processes != 0:
```
See my comment above about this requirement.

lvwerra reviewed (commented):
Hi @liyongsea 

Thank you so much for implementing this. I think this will be a great addition for everybody working on these models as evaluation is definitely a bottleneck at the moment - I am really excited about this feature!

I left a few comments. My main concern is about the current distribution of the data to multiple GPUs which requires the matching the number of generations with a multiple of the GPUs. I think we should remove that before for a better generalisability. E.g. if we would like to run a temperature sweep for the large CodeParrot model I would likely use a machine with 16 GPUs. Another use case is when we want to run AlphaCode evaluation with O(100k-1M) generations we would likely need to run it on even larger infra. Happy to help figuring out how to make it work with the padding across processes. I left a few suggestions how it might work. 

Another note is that I think we should replace `human_eval.py` with your script as your script also allows to evaluate on a single GPU and I would like to avoid unnecessary code duplication which would just make it more confusing for the user. We probably just need to add a note in the instruction readme about setting the number of workers.

liyongsea commented:
> There seems to be also a minor issue with the code style. The CI enforces some standards and you can fix any formatting issues with:
> 
> ```shell
> make fixup
> ```
> 
> For more information on both rebasing/merging and quality checks you can checkout the [contributing guide](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md).

Thank you for your feedback here is what I am going the go:
- Rebase and fix the formatting issue
- use task id to reorganize the output
- remove padding before generation to avoid the 'n_copies' tricks

Commit:
update code style

Commit:
use task id to arrange output

Commit:
truncate input to avoid zero pad

Commit:
Stop the copy mechanism

Commit:
update style

Commit:
restore copies to scale better in distributed mode

Commit:
update style

Commit:
replace human eval

liyongsea commented:
hi @lvwerra 
I finished the implementation. Please start review. Here are some area you might need to pay extra attention.
- I keep the behavior of sending copies of prompts, such that when generating 1M prediction per prompt, the same task is also distributed across worker
- Padding is removed before generation. However, the code still requires n_tasks x n_samples is a multiple of n_processes. Because the length of the dataloader should be a multiple of n_processes (do you know how to easily avoid this). Do you think it is ok? Or shall I change the behavior?

lvwerra reviewed (commented):
Hi @liyongsea 

Thanks for making the changes - I think it is much cleaner now! I left a few small comments and nits. The main change is about the tokenization of the prompts which should simplify the code even further. 

I think your first point is fine, that shouldn't be an issue. I am sure generation is the current bottleneck and not sending a few copies of the prompts. 

Regarding your second point: I don't think this is true - `accelerate` should be able to handle that! It just means that during the last batch of jobs not all GPUs will be utilized but I think that is fine. Did you run into any problems? If not we could remove the `ValueError`.

liyongsea committed:
Apply suggestions from code review

1. Tokenize all input at the same time
2. use attention_mask to get the input length
3. other small fixes

Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>

liyongsea commented:
@lvwerra Thank you for your suggestions. I learnt a lot ! It is integrated into the PR, I will test it.

> Regarding your second point: I don't think this is true - accelerate should be able to handle that! It just means that during the last batch of jobs not all GPUs will be utilized but I think that is fine. Did you run into any problems? If not we could remove the ValueError.

I have an error message when doing so, I will dig into it

Commit:
correct typo and update docstring

Commit:
update code style

Commit:
remove num sample division constraint

Commit:
remove max len calculation

liyongsea commented:
Hi @lvwerra I updated the code according to your suggestion. And I remove the exception regarding the division issue (please see my comment above).
I really appreciated your time. Tell me if you spot anything.

lvwerra commented on examples/research_projects/codeparrot/scripts/human_eval.py:
```diff
@@ -39,16 +70,78 @@ def __call__(self, input_ids, scores, **kwargs):
+    for step, batch in tqdm(enumerate(dataloader)):
+        with torch.no_grad():
+            # do not generate if the batch is empty, it could be the case when multi GPU is used
+            if len(batch["ids"]):
```
I am curious - is it actually empty sometimes? I don't remember needing to do any of that in the training script. I think the dataloader would just stop for that worker earlier and never enter the loop in the last iteration.

lvwerra commented on examples/research_projects/codeparrot/scripts/human_eval.py:
```diff
@@ -2,26 +2,57 @@
     StoppingCriteria,
     StoppingCriteriaList,
-    pipeline,
     set_seed,
```
It just occurred to me that we should probably switch to the `set_seed` function from `accelerate`:
https://github.com/huggingface/accelerate/blob/2c554b056cf10f7ae26e5d04cb60bf9dc35812e4/src/accelerate/utils.py#L120-L137
Otherwise what could happen is that each worker has the same seed and thus generates the same sequences even when sampling. We should set `device_specific=True` such that each worker gets a different seed.

liyongsea replied:
Good idea, I was wondering the same issue as well. I will change it

lvwerra reviewed (commented):
Awesome, we are close to the finish line. We'll make a test run on our side as well (cc @loubnabnl). 

In the meantime @muellerzr or @sgugger would you mind having a quick look, too? The goal is to move the old evaluation script to `accelerate` so we can make the time consuming generations faster with parallelization. In this script we need to generate _k_ code completions for each problem in `HumanEval` which are then evaluated at the end (that part is `accelerate` agnostic as it runs on the CPU). Let us know if we are using `accelerate` correctly for that task. Thanks!

muellerzr commented on examples/research_projects/codeparrot/scripts/human_eval.py:
```diff
@@ -39,16 +70,78 @@ def __call__(self, input_ids, scores, **kwargs):
+                generated_tokens, dim=1, pad_index=tokenizer.pad_token_id
+            )
+
+            generated_tokens = accelerator.gather(generated_tokens).cpu().numpy()
+            generated_tasks = accelerator.gather(generated_tasks).cpu().numpy()
```
You should be able to reduce this to:

```python
generated_tokens, generated_tasks = accelerator.gather(generated_tokens, generated_tasks)
generated_tokens = generated_tokens.cpu().numpy()
generated_tasks = generated_tasks.cpu().numpy()
```
And quick playing on my own machine reduced this by roughly half. (it's in microseconds, but it's a small quick efficiency boost)

sgugger replied:
Oh, interesting it gives a boost as it's not doing anything different then gathering the two in a list comprehension.

muellerzr replied:
Yeah, was surprised as well.

Here's what I toyed with:
```python
from accelerate import Accelerator
import torch

accelerator = Accelerator()
a = torch.tensor([1,2,3])
b = torch.tensor([[1],[2],[3]])
```
And then for timing:
```python
%%timeit -n 10
c,d = accelerator.gather((a,b))
c = c.cpu().numpy()
d = d.cpu().numpy()
```
```python
%%timeit -n 10
c = accelerator.gather(a).cpu().numpy()
d = accelerator.gather(b).cpu().numpy()
```
The first on Colab took 25µs as the best, while the latter took 48.9

liyongsea replied:
Thank you for sharing! I will do the update

muellerzr reviewed (commented):
It all looks pretty good to me! There was one small nit I found where we could speed things up slightly. 

(Also unsure of your empty tensor issue)

sgugger reviewed (approved):
LGTM as long as @lvwerra is happy :-)

liyongsea commented:
@lvwerra for the empty tensor issue, I found this in the documentation
https://github.com/huggingface/accelerate/blob/1d95ebdaa40ed42d0bb2b41ba5531b49eb8948b1/src/accelerate/data_loader.py#L204
I am really not sure if it is related. Tomorrow I will make a smaller example to verify this

Commit:
use accelerator.gather once to speed up

loubnabnl commented:
Hi @liyongsea , thank you for your contributions. I made a test run on my side and the code works fine!
For an NVIDIA Tesla A100 with 2 GPUs and the parameters you're using, I get the following results
|#processes | Time | pass@1 | pass@10| pass@100| 
|   -   |   -  |   -  |    -  |    -  | 
| 1 | 3:07:26 | 3.41% | 5.39% | 7.01% |
| 2 | 1:36:34 | 3.51% | 5.63% | 6.67% |

lvwerra commented:
So as the numbers seem to add up I think we can merge this once the `set_seed` is added. The check for empty batches is simple enough so we can leave it like that for now in my opinion - no reason to spend much time on that. Thanks again for adding this!

Commit:
use accelerate set_seed; update accelerate version

Commit:
correct gather bug

liyongsea commented:
> So as the numbers seem to add up I think we can merge this once the `set_seed` is added. The check for empty batches is simple enough so we can leave it like that for now in my opinion - no reason to spend much time on that. Thanks again for adding this!

Hi @lvwerra 
I have finished the modification.
- accelerate.utils.set_seed is used with device_specific=True
- I have updated the accelerated version which solve the empty tensor issue

Feel free to run the whole script again (I even found the performance increased ...) and merge if you are ok !
